### PR TITLE
runner constructor

### PR DESCRIPTION
**Pull request by densechen**
Add `RunnerConstructor` feature to build_runner.

It will not effect the original code feature, but just an enhancement.

Refer here for more details: [issue1225](https://github.com/open-mmlab/mmcv/issues/1225#issuecomment-891522362)

### CODE REVIEW

**Comment by densechen** [comment]
import in `__init__.py`

**Comment by ZwwWayne** [comment]
Hi @densechen ，
Thanks for the PR. Could you show the new use cases used in your projects after introducing `RunnerConstructor`? The information would be beneficial for us to consider the new design.

**Comment by densechen** [comment]
Hi, @ZwwWayne , if a `RunnerConstructor` is provided, we can have the ability to make some extra operations on the build runner without the necessary to change the user code. 
By the way, we have developed a new open source pytorch-style federated learning library, which will be released soon. 
We are currently palning to write an extra package that constumed for mm-xxx project. It will make all the mm-xxx project enable to do federated learning experiments without the necessary to change the user code, but just import our extra package at the beginning of code.
While developing this extra package, we found that it may be more flexible to provide a `RunnerConstructor` in our case.

**Comment by zhouzaida** [comment]
> Hi, @ZwwWayne , if a `RunnerConstructor` is provided, we can have the ability to make some extra operations on the build runner without the necessary to change the user code.
> By the way, we have developed a new open source pytorch-style federated learning library, which will be released soon.
> We are currently palning to write an extra package that constumed for mm-xxx project. It will make all the mm-xxx project enable to do federated learning experiments without the necessary to change the user code, but just import our extra package at the beginning of code.
> While developing this extra package, we found that it may be more flexible to provide a `RunnerConstructor` in our case.

Got it. Thanks for your contribution. And could you provide an example to show how to use the `RunnerConstructor`

**Comment by densechen** [comment]
Currently, if we want to do something with runner provided by mmcv, we need such piece of code: 

```python
from mmcv.runner.builder import RUNNERS, build_runner
from openfed.common.gluer import glue
from typing import Any

from .builder import build_openfed

def openfed_runner(model,
                   batch_processor=None,
                   optimizer=None,
                   work_dir=None,
                   logger=None,
                   meta=None,
                   max_iters=None,
                   max_epochs=None,
                   runner_cfg=None):
    # Build Runner
    runner = build_runner(
        runner_cfg,
        default_args=dict(
            model=model,
            batch_processor=batch_processor,
            optimizer=optimizer,
            work_dir=work_dir,
            logger=logger,
            meta=meta,
            max_iters=max_iters,
            max_epochs=max_epochs,
        )
    )
    # DO SOMETHING HERE
    runner.openfed_name = 'XXX'
    return runner


@RUNNERS.register_module(name='OpenFedRunner')
class OpenFedRunner(object):
    def __init__(self, *args, **kwargs):
        super().__init__()
        self.openfed_runner = openfed_runner(*args, **kwargs)

    def __getattribute__(self, name: str) -> Any:
        if name == 'openfed_runner':
            return super().__getattribute__(name)
        return getattr(self.openfed_runner, name)

    def __setattr__(self, name: str, value: Any) -> None:
        if name == 'openfed_runner':
            return super().__setattr__(name, value)
        return setattr(self.openfed_runner, name, value)
```

This code is quite ugly.
But if a `runner_constructor` is provided, we only need to define a new `runner_constructor` for doing this.
It seems more beautiful and flexible.
We are looking forward it.

**Comment by zhouzaida** [comment]
got it

**Comment by zhouzaida** [comment]
hi, please fix the CI

**Comment by densechen** [comment]
fix yapf

**Comment by zhouzaida** [comment]
> fix yapf

you can refer to https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md

**Comment by densechen** [comment]
fix yapf

**Comment by densechen** [comment]
Waiting for merging
@zhouzaida

**Comment by zhouzaida** [comment]
> Waiting for merging
> @zhouzaida

hi @densechen , thanks for your contribution again. Before merging, An unittest should be added (refer to https://github.com/open-mmlab/mmcv/blob/master/tests/test_runner/test_optimizer.py). In addition, maybe we can enhance the docstring of DefaultRunnerConstructor like `example` or `Args`

File: mmcv/runner/builder.py
Comment (ZwwWayne) [inline mmcv/runner/builder.py:10]: We can simply use `RUNNER_BUILDERS.build(cfg)`

File: mmcv/runner/default_constructor.py
Comment (ZwwWayne) [inline mmcv/runner/default_constructor.py:18]: Simply, we can simply use RUNNERS.build(self.runner_cfg, default_args=self.default_args)

**Comment by ZwwWayne** [comment]
Overall, this PR looks good to me. We need to resolve conflict before the merge.

File: mmcv/runner/default_constructor.py
Comment (densechen) [reply mmcv/runner/default_constructor.py:18]: Got it

File: mmcv/runner/builder.py
Comment (densechen) [reply mmcv/runner/builder.py:10]: Got it

**Comment by densechen** [comment]
Hi, @zhouzaida Can we have a further discussion via WeChat? My Email is densechen@foxmail.com

**Comment by zhouzaida** [comment]
> Hi, @zhouzaida Can we have a further discussion via WeChat? My Email is [densechen@foxmail.com](mailto:densechen@foxmail.com)

Great, I send an email to you

**Comment by ZwwWayne** [comment]
This PR can be merged after resolving linting issues and conflicts.

**Comment by densechen** [comment]
Using Examples: Inject some new properties and functions for existing `Runner`, such as `EpochBasedRunner`, though a `RunnerReconstructor`.

## 1. Define a new `RunnerReconstructor`:
```python
from mmcv.runner import RUNNER_BUILDERS

@RUNNER_BUILDERS.register_module()
class MyRunnerConstructor:
    def __init__(self, runner_cfg, default_args=None):
        if not isinstance(runner_cfg, dict):
            raise TypeError('runner_cfg should be a dict',
                            f'but got {type(runner_cfg)}')
        self.runner_cfg = runner_cfg
        self.default_args = default_args

    def __call__(self):
        runner = RUNNERS.build(self.runner_cfg, default_args=self.default_args)
        # Add new properties for existing runner
        runner.my_name = 'my_runner'
        runner.my_function = lambda self: print(self.my_name)
        ...
```

## 2. Use this `RunnerConstructor`, the config file is like:
```python
runner = dict(type='EpochBasedRunner', max_epochs=40, constructor='MyRunnerConstructor')
```

**Comment by ZwwWayne** [comment]
1. need to resolve conflicts.
2. Would you like to also update the usage of the runner in the documentation? You can also do that in a later PR.

File: mmcv/runner/default_constructor.py
Comment (ZwwWayne) [inline mmcv/runner/default_constructor.py:7]: [Nit]: update the docstring of this module

Comment (densechen) [reply mmcv/runner/default_constructor.py:7]: add examples. Thanks @zhouzaida

**Review by zhouzaida — APPROVED** [review]
LGTM

**Review by ZwwWayne — APPROVED** [review]

**Comment by densechen** [comment]
re-commit

**Comment by zhouzaida** [comment]
I am so sorry to force push to your master branch

**Comment by densechen** [comment]
recommit

**Comment by densechen** [comment]
https://github.com/open-mmlab/mmcv/pull/1296

**Comment by zhouzaida** [comment]
Thanks

**Pull request closed without merging**
